### PR TITLE
fix: adding backup if there are no active rollouts from the rollout dashboard

### PR DESCRIPTION
**Pull request by NikolaMilosa**
This PR adds a backup option in case there are no viable subnet Git revisions or the rollout dashboard is down. Before this PR the job would fail and say that the matrix was invalid which results in a failed release cut as we can see [here](https://dfinity.slack.com/archives/C024WJPPX5W/p1726110654204759). 

There are some occasions when we expect this situation with the current code base:

1. There was no successful cuts for at least a week.
2. We didn't release because of the holidays (Like we do for Christmas) 
3. We had an urgent fix that was rolled out in a day and no new rollouts were dispatched when the release-testing flow was issued.
4. Rollout dashboard is malfunctioning.

Yet another set of concerns need resolving.  The prior version selection algorithm (in the case of using the rollout dashboard) included all the versions of all the ongoing rollouts.  This has the following problems:

1. If there is no ongoing rollout, we obtain no version data.
2. If there are failed rollouts, but some subnets have in fact been rolled out successfully, we excluded those versions, even though quite logically those versions will be in mainnet if the failed rollout is the most recent rollout.
3. If there is a hotfix rollout, which only targets a handful of subnets, then versions may not be complete.

This PR proposes a way to only have a backup if everything is fine with our primary source of information (rollout dashboard) but there are no active rollouts **OR** if the rollout dashboard is unavailable. If its available but the data is malformed the CI will panic and will require investigation.

In addition to that we refine the algorithm used to select the versions from the primary source of information.  What we do now is:

* Go through time, from past to future, for each subnet, collecting each version that has been targeted to that
  subnet, along with its date (completion date if available, else start time if available, else planned start time) and state.
* If the state of that subnet's rollout instance is failed, then we disregard it.  Else, we take it into account.
* Having iterated through the rollouts one by one, we eventually end up with a list of subnets and a chronological list of
  revisions that have been targeted to those subnets (and either were successfully rolled out, or are pending and may be rolled out in the future).
* For each subnet, we pick (based on the date we collected) the most recent version.

Those are the versions that we're going to test against.

If this fails, that means that the versions on subnets of mainnet shall be the initial versions for qualification.

### CODE REVIEW

File: .github/scripts/determine-initial-guest-os-versions.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    """Try fetch data from `url`.
+
+    Will try to fetch data and return `None` if the resource is unavailable for some.
+    If this call fails other resources should be considered.
+    """
+    try:
+        req = Request(url)
+        req.add_header('accept', '*/*')
+        req.add_header('user-agent', 'dfinity-ci')
+        with urlopen(req, timeout=timeout) as response:
```
Comment (DFINITYManu) [inline .github/scripts/determine-initial-guest-os-versions.py:34]: Why not use `request.get()` from the requests library, and then `r.raise_for_status()` to handle non-200 responses?  urlopen?

File: .github/scripts/determine-initial-guest-os-versions.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    """
+    data = ensure_json_response(ROLLOUT_DASHBOARD_ENDPOINT)
+    if data is None:
+        # Resource was unavailable
+        return []
+
+    versions = set()
+    try:
+        for rollout in data:
+            if rollout.get(STATE, None) is None:
```
Comment (DFINITYManu) [inline .github/scripts/determine-initial-guest-os-versions.py:74]: This is unnecessary guard code.  The structure is not going to change from under you:

https://github.com/dfinity/dre-airflow/tree/main/rollout-dashboard/frontend/src/lib

File: .github/scripts/determine-initial-guest-os-versions.py (hunk 1 above)
Comment (NikolaMilosa) [reply .github/scripts/determine-initial-guest-os-versions.py:34]: This is native python, no need to waste time on installing pip requirements. Even the repo is sparse-checked out for only that file.

File: .github/scripts/determine-initial-guest-os-versions.py (hunk 2 above)
Comment (NikolaMilosa) [reply .github/scripts/determine-initial-guest-os-versions.py:74]: I'd rather it stay like this, to ensure everything is as it should be.

**Review by basvandijk — APPROVED** [review]

File: .github/scripts/determine-initial-guest-os-versions.py [hunk 3]
```diff
@@ -66,36 +70,54 @@ def fetch_versions_from_rollout_dashboard() -> list[str] | None:
...
 
-def maybe_exectued_timestamp(x : Any) -> int:
+    # The value of the dict entry is datestring, git revision.
+    subnet_to_revision: Dict[str, List[Tuple[str, str]]] = {}
+
+    for rollout in reversed(rollouts):  # Oldest to newest
+        for batch_num_ignored, batch in rollout["batches"].items():
+            for subnet in batch["subnets"]:
+                if subnet["state"] in (SubnetRolloutState.error, SubnetRolloutState.predecessor_failed):
+                    # This subnet failed?  We ignore it, because it could not have been upgraded.
```
Comment (DFINITYManu) [inline .github/scripts/determine-initial-guest-os-versions.py:100]: One refinement we could do here is to test whether a proposal to upgrade this subnet was filed.  This is not currently available in the rollout dashboard API but could be easily made available.

@NikolaMilosa what do you think?

File: .github/scripts/determine-initial-guest-os-versions.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    url = PUBLIC_DASHBOARD_ENDPOINT
+    try:
+        data = request_json(url)
+    except Exception as e:
+        eprint(f"Error fetching / decoding data from {url}: {e}.  Returning no versions.")
+        return []
+
+    versions = set()
+    # Manuel: I'm not modifying this because I do not know the shape
+    # of this data structure.
```
Comment (sasa-tomic) [inline .github/scripts/determine-initial-guest-os-versions.py:154]:
```suggestion
    # of this data structure.
    # Swagger for the public dashboard API: https://ic-api.internetcomputer.org/api/v3/swagger
```

Comment (DFINITYManu) [reply .github/scripts/determine-initial-guest-os-versions.py:154]: SWEG.

**Review by sasa-tomic — APPROVED** [review]
Looks solid!

File: .github/scripts/determine-initial-guest-os-versions.py (hunk 3 above)
Comment (NikolaMilosa) [reply .github/scripts/determine-initial-guest-os-versions.py:110]: I don't think thats needed but feel free to explore if you feel like it

File: .github/scripts/determine-initial-guest-os-versions.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    if not unique_versions:
+        eprint("No active rollouts found, will use versions from public dashboard")
+        unique_versions = fetch_versions_from_public_dashboard()
+
+    if not unique_versions:
+        # At this moment if we don't have any starting version we cannot proceed
+        raise RuntimeError(f"Didn't find any versions from:\n\t1. {ROLLOUT_DASHBOARD_ENDPOINT}\n\t2. {PUBLIC_DASHBOARD_ENDPOINT}")
+    eprint(f"Will qualify, starting from versions: {json.dumps(unique_versions)}")
+    matrix = {
+        "versions": unique_versions
```
Comment (NikolaMilosa) [inline .github/scripts/determine-initial-guest-os-versions.py:188]: A typo here. It should be `version`, not `versions`

Comment (DFINITYManu) [reply .github/scripts/determine-initial-guest-os-versions.py:188]: Oh.

File: .github/scripts/determine-initial-guest-os-versions.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+class PDSubnetsResponse(TypedDict):
+    subnets: List[PDSubnet]
+
+def eprint(*args, **kwargs):
+    print(*args, file=sys.stderr, **kwargs)
+
+def eprint_fmt(str, *args):
+    return # remove me to get some real action
+    print((str % args) if args else str, file=sys.stderr)
```
Comment (NikolaMilosa) [inline .github/scripts/determine-initial-guest-os-versions.py:77]: Why don't we just print all of that? I don't see why would we not want it to be as verbose as possible.

Comment (DFINITYManu) [reply .github/scripts/determine-initial-guest-os-versions.py:77]: It's thousands of lines.  Super noisy.

**Pull request merged**